### PR TITLE
[InstCombine] Canonicalize `icmp eq/ne (A ^ C), B` to `icmp eq/ne (A ^ B), C`

### LINKED ISSUE #65968: Simplification Comparison (Clang15 vs Clang16)
Consider the following two functions:
```cpp
#define I (-__INT_MAX__ - 1)

int foo(int x, int y)
{
  return ((x & I) ^ I) == ((~y & I) ^ I);
}

int bar(int x, int y)
{
  return (x ^ y) < 0;
}
```

Clang15 can simplify `foo` to `bar`, but Clang16 (and trunk) cannot.
https://godbolt.org/z/P3fMee9q3

Clang15
```asm
foo(int, int):                               # @foo(int, int)
        mov     eax, edi
        xor     eax, esi
        shr     eax, 31
        ret
bar(int, int):                               # @bar(int, int)
        mov     eax, edi
        xor     eax, esi
        shr     eax, 31
        ret
```

Clang16 (and trunk)
```asm
foo(int, int):                               # @foo(int, int)
        not     edi
        and     edi, -2147483648
        and     esi, -2147483648
        xor     eax, eax
        cmp     edi, esi
        sete    al
        ret
bar(int, int):                               # @bar(int, int)
        mov     eax, edi
        xor     eax, esi
        shr     eax, 31
        ret
```

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch canonicalizes `icmp eq/ne (A ^ Cst), B` to `icmp eq/ne (A ^ B), Cst` since the latter form exposes more optimizations.
Proof: https://alive2.llvm.org/ce/z/9DbhGc
Fixes #65968.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -5437,6 +5437,15 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
         Pred, A,
         Builder.CreateIntrinsic(Op0->getType(), Intrinsic::fshl, {A, A, B}));
 
+  // Canonicalize:
+  // icmp eq/ne OneUse(A ^ Cst), B --> icmp eq/ne (A ^ B), Cst
+  Constant *Cst;
+  if (match(&I,
+            m_c_ICmp(PredUnused, m_OneUse(m_Xor(m_Value(A), m_Constant(Cst))),
+                     m_Value(B))))
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5445]: Probably immconstant here.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -5437,6 +5437,15 @@ Instruction *InstCombinerImpl::foldICmpEquality(ICmpInst &I) {
...
         Builder.CreateIntrinsic(Op0->getType(), Intrinsic::fshl, {A, A, B}));
 
+  // Canonicalize:
+  // icmp eq/ne OneUse(A ^ Cst), B --> icmp eq/ne (A ^ B), Cst
+  Constant *Cst;
+  if (match(&I, m_c_ICmp(PredUnused,
+                         m_OneUse(m_Xor(m_Value(A), m_ImmConstant(Cst))),
+                         m_Value(B))))
+    return ICmpInst::Create(Instruction::ICmp, Pred, Builder.CreateXor(A, B),
+                            Cst);
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5447]: Use `new ICmpInst` instead.

File: llvm/test/Transforms/InstCombine/icmp-equality-xor.ll [hunk 3]
```diff
@@ -83,8 +83,8 @@ define i1 @cmpeq_xor_cst1_multiuse(i32 %a, i32 %b) {
 
 define i1 @cmpeq_xor_cst1_commuted(i32 %a, i32 %b) {
 ; CHECK-LABEL: @cmpeq_xor_cst1_commuted(
-; CHECK-NEXT:    [[C:%.*]] = xor i32 [[A:%.*]], 10
-; CHECK-NEXT:    [[CMP:%.*]] = icmp eq i32 [[C]], [[B:%.*]]
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/icmp-equality-xor.ll:87]: Not actually commuted. You need "thwart complexity-based canonicalization".

**Review by nikic — APPROVED** [review]
Sounds plausible, LGTM.

**Comment by dyung** [comment]
@dtcxzyw this change seems to be causing an infinite loop in the compiler in some of our tests. Can you take a look at #67783?

**Pull request merged**